persist-cli-tests/src/test/java/io/ballerina/persist/tools/ToolingDbPullTest.java

```diff
@@ -0,0 +1,447 @@
+    public void pullTestSelfReferencedRelation() throws BalException {
+        String subDir = "tool_test_pull_18_mysql";
+        createFromDatabaseScript(subDir, "mysql", databaseConfig);
+        executeDefaultPullCommand(subDir);
+        assertGeneratedSources(subDir);
+    }
+
+    @Test(enabled = true)
+    @Description("Create a model.bal file consisting of two relations which cross-reference each other.")
+    public void pullTestCrossReferencedRelations() throws BalException {
+        String subDir = "tool_test_pull_19_mysql";
+        createFromDatabaseScript(subDir, "mysql", databaseConfig);
+        executeDefaultPullCommand(subDir);
+        assertGeneratedSources(subDir);
+    }
+
+    @Test(enabled = true)
+    @Description("Create a model.bal file consisting keyword field names which are to be escaped.")
+    public void pullTestEscapedFieldNames() throws BalException {
+        String subDir = "tool_test_pull_20_mysql";
+        createFromDatabaseScript(subDir, "mysql", databaseConfig);
+        executeDefaultPullCommand(subDir);
+        assertGeneratedSources(subDir);
+    }
+
+    @Test(enabled = true)
+    @Description("When the database does not contain any tables.")
+    public void pullTestNoTablesInDatabase() throws BalException {
+        String subDir = "tool_test_pull_21_mysql";
+        createFromDatabaseScript(subDir, "mysql", databaseConfig);
+        executeDefaultPullCommand(subDir);
+        assertGeneratedSources(subDir);
+    }
+
+    @Test(enabled = true)
+    @Description("When the database does not exist.")
+    public void pullTestDatabaseDoesNotExist() throws BalException {
+        String subDir = "tool_test_pull_22_mysql";
+        createFromDatabaseScript(subDir, "mysql", databaseConfig);
+        executeDefaultPullCommand(subDir);
+        assertGeneratedSources(subDir);
+    }
+
+    private static void executeDefaultPullCommand(String subDir) throws BalException {
+        try {
+            Class<?> persistClass = Class.forName("io.ballerina.persist.cmd.Pull");
+            Pull persistCmd = (Pull) persistClass.getDeclaredConstructor(String.class).
+                    newInstance(Paths.get(GENERATED_SOURCES_DIRECTORY, subDir).toAbsolutePath().
+                            toString());
+            new CommandLine(persistCmd).parseArgs("--datastore", "mysql");
+            new CommandLine(persistCmd).parseArgs("--host", "localhost");
+            new CommandLine(persistCmd).parseArgs("--port", "3307");
+            new CommandLine(persistCmd).parseArgs("--user", "root");
+            new CommandLine(persistCmd).parseArgs("--database", "persist");
+            String password = "Test123#\n";
+            InputStream originalSystemIn = System.in;
+            InputStream inputStream = new ByteArrayInputStream(password.getBytes(StandardCharsets.UTF_8));
+            System.setIn(inputStream);
+            persistCmd.execute();
+            System.setIn(originalSystemIn);
```

Comment:
Shall we have this inputStream in a try block for proper resource management?
```suggestion
try (InputStream inputStream = new ByteArrayInputStream(password.getBytes(StandardCharsets.UTF_8))) {
        System.setIn(inputStream);
        // Execute the command
        persistCmd.execute();
    } finally {
        // Restore the original System.in
        System.setIn(originalSystemIn);
    }
```
Will the above work?